internal/storage/bucket_handle_test.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2022 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"context"
+	"testing"
+
+	"github.com/fsouza/fake-gcs-server/fakestorage"
+	"github.com/jacobsa/gcloud/gcs"
+	. "github.com/jacobsa/ogletest"
+)
+
+const missingObjectName string = "test/foo"
+
+func TestBucketHandle(t *testing.T) { RunTests(t) }
+
+type BucketHandleTest struct {
+	fakeStorageServer *fakestorage.Server
+	bucketHandle      *bucketHandle
+}
+
+var _ SetUpInterface = &BucketHandleTest{}
+var _ TearDownInterface = &BucketHandleTest{}
+
+func init() { RegisterTestSuite(&BucketHandleTest{}) }
+
+func (t *BucketHandleTest) SetUp(_ *TestInfo) {
+	var err error
+	t.fakeStorageServer, err = CreateFakeStorageServer([]fakestorage.Object{GetDefaultObject()})
+	AssertEq(nil, err)
+
+	storageClient := &storageClient{client: t.fakeStorageServer.Client()}
+	t.bucketHandle, err = storageClient.BucketHandle(DefaultBucketName)
+	AssertEq(nil, err)
+	AssertNe(nil, t.bucketHandle)
+}
+
+func (t *BucketHandleTest) TearDown() {
+	t.fakeStorageServer.Stop()
+}
+
+func (t *BucketHandleTest) TestNewReaderMethodWithCompleteRead() {
+	rc, err := t.bucketHandle.NewReader(context.Background(),
+		&gcs.ReadObjectRequest{
+			Name: DefaultObjectName,
+			Range: &gcs.ByteRange{
+				Start: uint64(0),
+				Limit: uint64(len(ContentInDefaultObject)),
+			},
+		})
+
+	AssertEq(nil, err)
+	buf := make([]byte, len(ContentInDefaultObject))
+	_, err = rc.Read(buf)
+	AssertEq(nil, err)
+	ExpectEq(string(buf[:]), ContentInDefaultObject)
+}
+
+func (t *BucketHandleTest) TestNewReaderMethodWithRangeRead() {
+	start := uint64(2)
+	limit := uint64(8)
+
+	rc, err := t.bucketHandle.NewReader(context.Background(),
+		&gcs.ReadObjectRequest{
+			Name: DefaultObjectName,
+			Range: &gcs.ByteRange{
+				Start: start,
+				Limit: limit,
+			},
+		})
+
+	AssertEq(nil, err)
+	buf := make([]byte, limit-start)
+	_, err = rc.Read(buf)
+	AssertEq(nil, err)
+	ExpectEq(string(buf[:]), ContentInDefaultObject[start:limit])
+}
+
+func (t *BucketHandleTest) TestNewReaderMethodWithInValidObject() {
+	rc, err := t.bucketHandle.NewReader(context.Background(),
+		&gcs.ReadObjectRequest{
+			Name: missingObjectName,
+			Range: &gcs.ByteRange{
+				Start: uint64(0),
+				Limit: uint64(len(ContentInDefaultObject)),
+			},
+		})
+
+	AssertNe(nil, err)
+	AssertEq(nil, rc)
+}
+
+func (t *BucketHandleTest) TestNewReaderMethodWithGeneration() {
```

Comment:
We need to enable the bucket versioning for this test.

Reply:
I didnt follow. what is pending here?

Reply:
Object versioning related test, but we are able to read the object with specific version. Right now, older object gets deleted when we modify the object.

Reply:
Added the comment on the top of the file.